File: go/expr/expr_test.go
package expr

import (
	"fmt"
	"math/rand/v2"
	"testing"
)

var random = rand.New(rand.NewPCG(0, 0))

func TestRandomCases(t *testing.T) {
	for test := 0; test < 100_000; test++ {
		expr, want, _ := genExpr(random.IntN(10))
		got := EvaluateExpr(expr)
		if got != want {
			t.Fatalf("EvaluateExpr(%v):\nwant %v\n got %v", expr, want, got)
		}
	}
}

func genExpr(depth int) (expr string, val int, add bool) {
	if depth == 0 {
		v := random.IntN(10)
		return fmt.Sprint(v), v, false
	}

	n := 1 + random.IntN(5)
	add = random.IntN(2) == 0

	for i := 0; i < n; i++ {
		e, v, a := genExpr(random.IntN(depth))
		if a && !add || random.IntN(10) == 0 {
			e = "(" + e + ")"
		}

		if i == 0 {
			expr, val = e, v
			continue
		}

		if op := random.IntN(2); add {
			if op == 0 {
				expr += "+" + e
				val += v
			} else {
				expr += "-(" + e + ")"
				val -= v
			}
		} else {
			if op == 0 {
				expr += "*(" + e + ")"
				val *= v
			} else {
				expr += "/(" + e + ")"
				if v != 0 {
					val /= v
				} else {
					val = 0
				}
			}
		}
	}

	return expr, val, add
}

File: python/expr/expr_test.py
import random as rand
from expr import evaluate_expr

random = rand.Random(0)


def test_example():
	assert evaluate_expr('((1+2)+3*4-5*(6+7))/(0-5)') == 10


def test_random_cases():
	for _ in range(10_000):
		expr, want, _ = gen_expr(random.randrange(10))
		got = evaluate_expr(expr)
		assert got == want, \
			f'evaluate_expr({expr}):\nwant {want}\n got {got}'


def gen_expr(depth):
	expr, val, add = '', 0, False

	if depth == 0:
		v = random.randrange(10)
		return f'{v}', v, False

	n = 1 + random.randrange(5)
	add = random.randrange(2) == 0

	for i in range(n):
		e, v, a = gen_expr(random.randrange(depth))
		if a and not add or random.randrange(10) == 0:
			e = '(' + e + ')'

		if i == 0:
			expr, val = e, v
			continue

		op = random.randrange(2)
		if add:
			if op == 0:
				expr += '+' + e
				val += v
			else:
				expr += '-(' + e + ')'
				val -= v
		else:
			if op == 0:
				expr += '*(' + e + ')'
				val *= v
			else:
				expr += '/(' + e + ')'
				if v != 0:
					val //= v
				else:
					val = 0

	return expr, val, add
